keps/sig-node/3721-support-for-env-files/README.md

```diff
@@ -0,0 +1,470 @@
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This is a proposal to add capability to populate environment variables in the
+container from a file.
+
+## Motivation
+
+Kubernetes provides existing mechanisms for setting environment variables within
+containers using ConfigMap and Secret resources. However, these methods
+necessitate additional API calls. If a user wishes to populate environment
+variables directly from a file, a workaround requiring the execution of the
+`source <envfile>` command within the primary container is necessary. This
+approach is unsuitable in cases where the user cannot modify the container image
+or execute commands within it.
+
+The syntactic enhancement proposed in this KEP offers a simplified method for
+populating environment variables directly from files. This change would
+streamline the process for users.
+
+### Goals
+
+- Define container environment variables from a file.
+
+## Proposal
+
+### User Stories (Optional)
+
+#### Story 1
+
+The user wants to define container environment variables from a file. It can
+generate an env file using an `initContainer` or mount from the host. The
+container can then use the env file to define its environment variables. In the
+example below, the Pod's output includes `CONFIG_VAR=HELLO`.
+
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  name: test-pod
+spec:
+  initContainers:
+  - name: setup-envfile
+    image: registry.k8s.io/busybox
+    command: ['sh', '-c', 'echo CONFIG_VAR="HELLO" > /config.env']
+    volumeMounts:
+    - name: data
+      mountPath: /
+  containers:
+  - name: use-envfile
+    image: registry.k8s.io/busybox
+    command: [ "/bin/sh", "-c", "env" ]
+    envFrom:
+    - fileRef:
+        path: /config.env
```

Comment:
Ideally, it is not needed.

I feel like there are a few ways a user can use this feature. One of the ways would be an `initContainer` that generates the env file in a volume. This volume can be an emptyDir/pvc/hostPath.

When the primary container uses this generated env file, it can just specify the host path of that file. But that host path would be difficult for an app developer to locate if it was originally created in an emptyDir or pvc by the `initContainer`. for eg, `emptyDir` volumes are by default under `/var/lib/kubelet/pods/PODUID/volumes/kubernetes.io~empty-dir/VOLUMENAME`. This would be too difficult to reference in a Deployment spec as UIDs will be different for different pods.

WDYT?

Reply:
I'd like the kubelet to be able to mount a volume, somewhere outside of all containers, and copy the environment from that mount. You'd specify what path inside the filesystem that the kubelet should parse.

Once the first app container has started, the kubelet doesn't need to keep that mount around.

Reply:
I modified the API to add a subfield `volumeName` to `fileRef` and `fileKeyRef` fields. This would allow kubelet to locate the env file and not keep the mount around when the container has started.